### PR TITLE
Added implementation and unit tests for ipv4 and ipv6 validation

### LINKED ISSUE #421: ipv4 validation not implemented
### General information
The ipv4 format specifies that the entered string should be in the format of an ipv4 address. I believe this was already specified in json schema v4 but doesn't seem to be implemented in json-editor.

- json-editor version: 1.4.0-beta.0
### Expected behavior

Should create a validation error if a non-valid ip address is entered into a string field specified with format ipv4.

### Actual behavior

Does not raise a validation error.

### Steps to reproduce the behavior

[Direct link to example](https://json-editor.github.io/json-editor/?schema=N4IgLglmA2CmIC4QCsDOB7AdgAlQYwAtYBbAQ21gA9TiAHOEAGnAE9b4l0AjZWPMJiFoAndO2GRYqRKAi0AggBNFwqdISgwbDiFRhhETAHNBAM3TCyApHIBuAFkGKpeA7UhZEIAJIAFbKTKqqioAHQgAL4RzKoAjgCuEKqKiADaIHJKKmogALoRQA===&value=N4IglgDgggJjBOBTAzskAuEAjANgQwAsQBfIA===&lib_switcher=&prompt_before_delete&upload=function(a,b,c){console.log(%22Upload%20handler%20required%20for%20upload%20editor%22)}&theme=bootstrap2&iconlib=fontawesome4&object_layout=normal&show_errors=interaction)

```json
{
  "title": "json schema example",
  "type": "object",
  "properties": {
    "ipAddress": {
      "type": "string",
      "format": "ipv4",
      "description": "IP address."
    }
  },
  "required": [
    "ipAddress"
  ]
}
```

### PR DESCRIPTION
**Pull request by sirockin**
Fixed #421.

ipv6 validator has (disabled) unit tests for but does not yet implement compressed-zero format.

### CODE REVIEW

File: src/validators/ip-validator.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Implements ipv4 and ipv6 format validations as perhttps://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-7.3.4
+var ipValidator = {
```
Comment (pmk65) [inline src/validators/ip-validator.js]: Shouldn't the validation code be included directly in` src/validator.js` rather than an external file?

Comment (sirockin) [reply src/validators/ip-validator.js]: Hi @pmk65, thanks for the feedback. Obviously it's good practice to break code down into small functional units since it makes the code more readable and testable. When and how you choose to distribute this is a matter of preferred coding style. I took my lead from the layout of the editor code, which has a single file then a subdirectory for specific implementations. I thought this approach would make it easier for future contributors to add extra validators where necessary, but it could equally be moved back into the main file.

There's one other thing to consider though, the mechanics of unit testing: Currently the code is built by simply concatenating files according to the order specified in the gruntfile. This is obviously a nice simple approach but does make it harder to manage dependencies and also to import files for testing: When unit testing you usually `require()` or `import` the files and hence the objects you want to test but you can't do this easily with the current setup since the 'Validator` class has external dependencies not declared in `validator.js`. So I can't simply export then import the `Validator` class - I have to export a standalone function using the  `if( typeof(exports)...` construct you see in `ip-validator.js`. And that seemed to go better in a different file.

tldr: It's a pain trying to unit-test code if it isn't built using a module-loader like requirejs or using `import` and `export` statements provided in later javascript versions.

Comment (pmk65) [reply src/validators/ip-validator.js]: @sirockin Placing the validators in separate files is not a bad idea. The same should also be done for the resolver codes in `src/default.js`. That way you can easily build a custom version with only the editors you need/want. But currently the editors needs an overhaul first #401 

As for the unit tests, are you using the **src/** files for the tests? Would it help using the **dist/** files? (which is a single file build from the files in` src/`) 

Is it possible to create a module-loader using the existing code base?

Comment (sirockin) [reply src/validators/ip-validator.js]: Hi @pmk65 . Using the dist build may work but I wouldn't advise it since it's better to automatically run unit tests on the source code before building. (for this repo we could do it from the gruntfile) 

I'm sure the existing code base could be fairly easily adapted to use require(). You may instead want to consider upgrading the javascript version to one which allows use of import/export as I think this is the preferred way of doing things these days. Again, this shouldn't be too difficult as all the javascript versions are backwards-compatible.

Comment (sirockin) [reply src/validators/ip-validator.js]: Actually `require()` is probably the best way to go (using [browserify](http://browserify.org/)) since javascript ES6 (and hence import) isn't yet supported in Internet Explorer.

**Comment by schmunk42** [comment]
@sirockin Do we need Karma only for debugging the unit tests? Does it the same job as CodeceptJS or could it be done with CodeceptJS? 

I guess I'd need to give this PR a try locally, since I wonder how you'd debug unit tests in a browser?!

**Comment by sirockin** [comment]
Essentially yes - the main test framework is Jasmine and you can run the unit tests directly from the command line using the Jasmine CLI. (see `test.unit` script in `package.json`).

Karma is a browser-based test runner. I'm not an expert but I've been able to debug in the browser by starting the `karma` `package.json` script, hitting the "Debug' button on the browser, then using the browser dev tools to view the console, set breakpoints etc.

**Comment by sirockin** [comment]
Sorry, regarding karma, it looks like the setup I gave you wasn't as useful as it could have been. Replacing the contents of karma.conf.js with what's below  gives a much more useful output via Chrome. 
```javascript
'use strict';

module.exports = function(karma) {
  karma.set({

    frameworks: [ 'jasmine', 'browserify' ],

    files: [
      'tests/unit/**/*.spec.js'
    ],

    plugins: [
      require('karma-jasmine'), 
      require('karma-browserify'),
      require('karma-chrome-launcher'), 
      require('karma-jasmine-html-reporter')
    ],

    client: {
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },

    reporters: ['progress','kjhtml'],


    preprocessors: {
      'tests/unit/**/*.spec.js': [ 'browserify' ]
    },

    browsers: [ 'Chrome' ],

    customLaunchers: {
      Chrome_with_debugging: {
      base: 'Chrome',
      flags: ['--remote-debugging-port=9222'],
      debug: true
      }
      },

    // browserify configuration
    browserify: {
      debug: true
    }
  });
};
```


This probably isn't the right way to update a PR so sorry about that!

**Comment by schmunk42** [comment]
Thank you!

**Comment by pmk65** [comment]
@schmunk42 I think this conflicts with #426 as both adds IPv4 and IPv6 validation (#426 adds the validation through the editor)
